### ISSUE TITLE
A bug of torch.arange for long sequence with fp16 data type

### ISSUE THREAD

**Issue by renll** [pytorch/pytorch#81926]
### 🐛 Describe the bug

The torch.arange function has a bug for long sequences with the data type of fp16.

The function works normally with a sequence length of 2048:
```
import torch
a= torch.tensor([1]).cuda()
print(torch.arange(2048, device = a.device, dtype=torch.half).long())
print(torch.arange(2048, device = a.device, dtype=torch.long))
```
Output: 
```
tensor([   0,    1,    2,  ..., 2045, 2046, 2047], device='cuda:0')
tensor([   0,    1,    2,  ..., 2045, 2046, 2047], device='cuda:0')
```

However, when the length is 4096:
```
import torch
a= torch.tensor([1]).cuda()
print(torch.arange(4096, device = a.device, dtype=torch.half).long())
print(torch.arange(4096, device = a.device, dtype=torch.long))
```
Output: 
```
tensor([   0,    1,    2,  ..., 4092, 4094, 4096], device='cuda:0') #Not expected!
tensor([   0,    1,    2,  ..., 4093, 4094, 4095], device='cuda:0')
```

This will cause some specific implementations of self-attention masking to leak future tokens (e.g. [Local Attention](https://github.com/lucidrains/local-attention)) under the long sequence scenario.

### Versions

PyTorch version: 1.12.0
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.6 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: version 3.10.2
Libc version: glibc-2.27

Python version: 3.8.13 (default, Mar 28 2022, 11:38:47)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.4.0-99-generic-x86_64-with-glibc2.17
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration:
GPU 0: NVIDIA GeForce GTX 1080 Ti
GPU 1: NVIDIA GeForce GTX 1080 Ti
GPU 2: NVIDIA GeForce GTX 1080 Ti
GPU 3: NVIDIA GeForce GTX 1080 Ti
GPU 4: NVIDIA GeForce GTX 1080 Ti
GPU 5: NVIDIA GeForce GTX 1080 Ti
GPU 6: NVIDIA GeForce GTX 1080 Ti
GPU 7: NVIDIA GeForce GTX 1080 Ti

Nvidia driver version: 470.103.01
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.22.3
[pip3] torch==1.12.0
[pip3] torchaudio==0.12.0
[pip3] torchvision==0.13.0
[conda] blas                      1.0                         mkl
[conda] cudatoolkit               10.2.89              hfd86e86_1
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] mkl                       2021.4.0           h06a4308_640
[conda] mkl-service               2.4.0            py38h7f8727e_0
[conda] mkl_fft                   1.3.1            py38hd3c417c_0
[conda] mkl_random                1.2.2            py38h51133e4_0
[conda] numpy                     1.22.3           py38he7a7128_0
[conda] numpy-base                1.22.3           py38hf524024_0
[conda] pytorch                   1.12.0          py3.8_cuda10.2_cudnn7.6.5_0    pytorch
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torchaudio                0.12.0               py38_cu102    pytorch
[conda] torchvision               0.13.0               py38_cu102    pytorch

**Comment by Gamrix** [comment]
This is due to the limitations of half precision numbers and not necessarily a pytorch bug. Half precision numbers can't odd numbers greater than 2048, and therefore arange() over that threshold is not defined behavior (though we might want to error on this case). 

Read precision limits here: 
https://en.wikipedia.org/wiki/Half-precision_floating-point_format

**Comment by renll** [comment]
That's not true. The largest possible value for fp16 is 65519 as said in the section you quoted.

> This is due to the limitations of half precision numbers and not necessarily a pytorch bug. Half precision numbers can't odd numbers greater than 2048, and therefore arange() over that threshold is not defined behavior (though we might want to error on this case).
> 
> Read the Precision Limits section of this Wikipedia article for more details. https://en.wikipedia.org/wiki/Half-precision_floating-point_format

**Comment by renll** [comment]
Yes, it makes sense, but considering the maximal value for fp16 is commonly believed as 65519, this kind of limitation of 2048 should be alerted.

**Comment by ngimel** [comment]
Closing as this is expected behavior.

**Comment by renll** [comment]
> Closing as this is expected behavior.

This is not the expected behavior. At least, you guys should clearly write in the doc that "this function does not support an input that is larger than 2048 under fp16, even though the fp16 tensor can support a maximal value of 65504".
